File: scripts/asset-pipeline/preview-v2.py
"""Render disposable authoring previews; never publish or select a production build.

Use Blender --background --factory-startup --python-exit-code 1 --python this.py
-- --repo <repo> --asset <id> --label <new label> [--indices 0,3,6,9].
"""
import argparse
import json
import os
from pathlib import Path
import sys
import bpy

sys.path.insert(0, str(Path(__file__).resolve().parent))
import blender_pipeline_v2 as pipeline


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', required=True)
    parser.add_argument('--asset', required=True)
    parser.add_argument('--label', required=True)
    parser.add_argument('--indices', default='0,3,6,9')
    parser.add_argument('--variant', choices=['calm', 'rich'], default='rich')
    parser.add_argument('--patch')
    parser.add_argument('--device', choices=['AUTO', 'CPU', 'CUDA', 'OPTIX'], default='AUTO')
    args = parser.parse_args(sys.argv[sys.argv.index('--')+1:])
    root = Path(args.repo).resolve()
    asset_id, label = pipeline.identifier(args.asset), pipeline.identifier(args.label)
    output = root / 'art/poc/pipeline-v2/previews' / label / asset_id
    if output.exists(): raise FileExistsError('Preview already exists; use a new label')
    catalog = json.loads((root / 'scripts/asset-pipeline/catalog-v2.json').read_text(encoding='utf-8'))
    spec = next(asset for asset in catalog['assets'] if asset['id'] == asset_id)
    if args.patch:
        patch = json.loads(Path(args.patch).read_text(encoding='utf-8-sig'))
        updates = patch.get('assets', patch) if isinstance(patch, dict) else patch
        update = updates[asset_id] if isinstance(updates, dict) else next(item for item in updates if item['id'] == asset_id)
        spec = spec | update
    # Probe in this disposable process; never change an interactive Blender session.
    device = 'CPU'
    prefs = bpy.context.preferences.addons['cycles'].preferences
    for candidate in (['OPTIX', 'CUDA', 'CPU'] if args.device == 'AUTO' else [args.device]):
        if candidate == 'CPU': break
        try:
            prefs.compute_device_type = candidate
            prefs.refresh_devices()
            if any(item.type == candidate for item in prefs.devices):
                device = candidate
                break
        except (TypeError, RuntimeError):
            pass
    if args.device not in ('AUTO', 'CPU') and device == 'CPU':
        raise ValueError(f'Requested GPU unavailable: {args.device}')
    fingerprint, sources, textures = pipeline.inputs(root, spec, device)
    session = pipeline.prepare(root, spec, label, fingerprint, sources, textures)
    scene, ctx = session['scene'], session['ctx']
    bpy.context.window.scene = scene
    if device != 'CPU':
        cache = root / 'art/poc/pipeline-v2/cache/optix'
        cache.mkdir(parents=True, exist_ok=True)
        os.environ['OPTIX_CACHE_PATH'] = str(cache)
        for item in prefs.devices: item.use = item.type == device
        scene.cycles.device = 'GPU'
    else:
        scene.cycles.device = 'CPU'
    settings = spec['materialVariants'][args.variant]
    for socket in ctx.strengths: socket.default_value = settings['textureStrength']
    for socket in ctx.form_strengths: socket.default_value = settings['formShadowStrength']
    indices = list(range(len(session['samples']))) if args.indices == 'all' else [int(value) for value in args.indices.split(',')]
    if not indices or any(index < 0 or index >= len(session['samples']) for index in indices): raise ValueError('Invalid preview frame')
    output.mkdir(parents=True)
    for index in indices:
        pipeline.set_frame(scene, session['samples'][index]['blenderFrame'])
        scene.render.filepath = str(output / f'frame-{index:04d}.png')
        bpy.ops.render.render(write_still=True)
    scene.frame_set(0)
    bpy.data.libraries.write(str(output / 'preview.blend'), {scene, *session['texts']}, fake_user=True)
    poses = []
    for index in indices:
        pose = dict(session['samples'][index], clip='idle', phase=0)
        for clip in session['clips']:
            if index in clip['frames']:
                position = clip['frames'].index(index)
                divisor = len(clip['frames']) if clip['loop'] else max(1, len(clip['frames']) - 1)
                pose.update(clip=clip['name'], phase=position / divisor)
        poses.append(pose)
    pipeline.save_json(output / 'preview.json', dict(status='authoring-preview', id=asset_id, spec=spec,
                       indices=indices, variant=args.variant, inputHash=fingerprint, bounds=session['bounds'],
                       baseDiameters=session['baseDiameters'], sources=sources, device=device,
                       poses=poses, clips=session['clips']))
    print(json.dumps({'preview': str(output), 'frames': indices}), flush=True)


if __name__ == '__main__': main()

File: scripts/asset-pipeline/preview-v2.mjs
// Disposable authoring only: production build/select/archive/import stay separate.
import { spawn } from 'node:child_process';
import { open, mkdir, readFile, writeFile, access } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { parseArgs } from 'node:util';
import { randomUUID } from 'node:crypto';
import { reviewPreview, safePath } from './review-preview-v2.mjs';

export const repoRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');
const identifier = value => {
  if (!/^[a-z0-9][a-z0-9-]*$/.test(value ?? '')) throw new Error('Asset/Label: Kleinbuchstaben, Zahlen und Bindestriche verwenden.');
  return value;
};

export function previewIndices(spec, selection) {
  const clips = spec.clips ?? [];
  if (clips.some(c => !Number.isInteger(c.frameCount) || c.frameCount < 1)) throw new Error('Ungültige Clip-Länge.');
  const count = 1 + clips.reduce((n, c) => n + c.frameCount, 0);
  const first = clips[0];
  const indices = selection === 'all' ? Array.from({ length: count }, (_, i) => i)
    : selection === undefined ? (first ? [0, ...[.25, .75].map(p => 1 + Math.min(first.frameCount - 1, Math.round(p * (first.loop ? first.frameCount : first.frameCount - 1))))] : [0])
    : /^\d+(,\d+)*$/.test(selection) ? selection.split(',').map(Number) : [];
  if (!indices.length || indices.some(i => i < 0 || i >= count)) throw new Error(`Ungültige Posen; erlaubt: 0–${count - 1} oder all.`);
  // Idle is needed for comparison even when the explicit selection omits it.
  return [...new Set([0, ...indices])];
}

export async function resolveBlender(explicit, env = process.env) {
  const configured = explicit || env.BLENDER_PATH;
  const executable = process.platform === 'win32' ? 'blender.exe' : 'blender';
  const candidates = configured ? [path.resolve(configured)] : (env.PATH ?? '').split(path.delimiter).filter(Boolean).map(p => path.join(p.replace(/^"|"$/g, ''), executable));
  for (const candidate of candidates) if (await access(candidate).then(() => true, () => false)) return candidate;
  throw new Error('Blender nicht gefunden. Aufruf: npm run assets:preview -- <id> --blender "<Pfad zur Blender-Programmdatei>" (alternativ BLENDER_PATH oder PATH).');
}

export async function renderProcess(executable, args, log, cwd) {
  const handle = await open(log, 'wx');
  try {
    await new Promise((resolve, reject) => {
      const child = spawn(executable, args, { cwd, windowsHide: true, stdio: ['ignore', handle.fd, handle.fd] });
      child.once('error', reject);
      child.once('close', (code, signal) => code === 0 ? resolve() : reject(new Error(`Blender fehlgeschlagen (Exit ${code ?? signal}).`)));
    });
  } catch (error) {
    const tail = (await readFile(log, 'utf8')).trim().split(/\r?\n/).filter(Boolean).slice(-12).join('\n');
    throw new Error(`${error.message}\n${tail}\nLog: ${log}`);
  } finally { await handle.close(); }
}

export async function preview(argv, root = repoRoot) {
  const { values: args, positionals } = parseArgs({ args: argv, allowPositionals: true, options: Object.fromEntries(
    ['blender', 'label', 'variant', 'indices', 'device', 'patch', 'compare', 'review'].map(k => [k, { type: 'string' }]).concat([['help', { type: 'boolean' }]])) });
  if (args.help) {
    console.log('npm run assets:preview -- <id> [--variant calm|rich] [--indices 0,4,10|all] [--label name] [--device AUTO|OPTIX|CUDA|CPU] [--blender path] [--patch path] [--compare image]\nOhne Blender erneut auswerten: npm run assets:preview -- --review <Vorschauordner> [--compare image]');
    return;
  }
  if (args.review) {
    if (positionals.length || Object.keys(args).some(k => !['review', 'compare'].includes(k))) throw new Error('--review nur mit --compare kombinieren.');
    return reviewPreview(path.resolve(root, args.review), { root, compare: args.compare });
  }
  if (positionals.length !== 1) throw new Error('Eine Asset-ID angeben; siehe --help.');
  const id = identifier(positionals[0]);
  const catalog = JSON.parse(await readFile(path.join(root, 'scripts/asset-pipeline/catalog-v2.json'), 'utf8'));
  let spec = catalog.assets.find(a => a.id === id);
  if (!spec || spec.production === 'planned') throw new Error(`Keine ausführbare Rezeptur: ${id}`);
  let patch;
  if (args.patch) {
    patch = path.resolve(root, args.patch);
    const parsed = JSON.parse((await readFile(patch, 'utf8')).replace(/^\uFEFF/, ''));
    const updates = parsed.assets ?? parsed;
    const update = Array.isArray(updates) ? updates.find(a => a.id === id) : updates[id];
    if (!update) throw new Error(`Patch enthält ${id} nicht.`);
    if (update.id && update.id !== id) throw new Error('Patch darf die Asset-ID nicht ändern.');
    spec = { ...spec, ...update };
  }
  const indices = previewIndices(spec, args.indices);
  const variant = args.variant ?? 'rich', device = args.device ?? 'AUTO';
  if (!['calm', 'rich'].includes(variant) || !['AUTO', 'CPU', 'CUDA', 'OPTIX'].includes(device)) throw new Error('Ungültige Variante oder Gerät; siehe --help.');
  const label = identifier(args.label ?? `preview-${new Date().toISOString().replace(/[^0-9]/g, '')}-${randomUUID().slice(0, 8)}`);
  const previews = await safePath(root, 'art/poc/pipeline-v2/previews');
  const folder = await safePath(previews, `${label}/${id}`);
  if (await access(path.dirname(folder)).then(() => true, () => false)) throw new Error('Preview-Label existiert bereits; neues Label verwenden.');
  const blender = await resolveBlender(args.blender);
  await mkdir(path.dirname(folder), { recursive: true });
  const log = path.join(path.dirname(folder), 'blender.log');
  const started = Date.now();
  await renderProcess(blender, ['--background', '--factory-startup', '--python-exit-code', '1', '--python',
    path.join(root, 'scripts/asset-pipeline/preview-v2.py'), '--', '--repo', root, '--asset', id, '--label', label,
    '--variant', variant, '--indices', indices.join(','), '--device', device, ...(patch ? ['--patch', patch] : [])], log, root);
  const manifestFile = path.join(folder, 'preview.json');
  const manifest = JSON.parse(await readFile(manifestFile, 'utf8'));
  manifest.durationSeconds = Math.round((Date.now() - started) / 100) / 10;
  manifest.log = path.relative(root, log).replaceAll('\\', '/');
  await writeFile(manifestFile, JSON.stringify(manifest, null, 2) + '\n');
  return reviewPreview(folder, { root, compare: args.compare });
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  preview(process.argv.slice(2)).catch(error => { console.error(error.message); process.exitCode = 1; });
}
